### DISCUSSION TITLE
snap_getBip44Entropy command

### DISCUSSION THREAD

**Discussion by MrDonaldBean** [MetaMask/snaps#985] [Q&A]
Hi, I'm a really noob user.

I'd like to use the snap_getBip44Entropy command...I've run correctly the "confirmation screen" example of the "getting started" tutorial and I've installed @metamaks/key-tree.

Then I've copied:
------------------------------------------
{
  "initialPermissions": {
    "snap_getBip44Entropy": [
      {
        "coinType": 3
      }
    ]
  }
}
----------------------------------------

in snap.manifest.json and:
------------------------------------------
import { getBIP44AddressKeyDeriver } from '@metamask/key-tree';

// By way of example, we will use Dogecoin, which has `coin_type` 3.
const dogecoinNode = await wallet.request({
  method: 'snap_getBip44Entropy',
  params: {
    coinType: 3,
  },
});

// Next, we'll create an address key deriver function for the Dogecoin coin_type node.
// In this case, its path will be: m / 44' / 3' / 0' / 0 / address_index
const deriveDogecoinAddress = await getBIP44AddressKeyDeriver(dogecoinNode);

// These are BIP-44 nodes containing the extended private keys for
// the respective derivation paths.

// m / 44' / 3' / 0' / 0 / 0
const addressKey0 = await deriveDogecoinAddress(0);

// m / 44' / 3' / 0' / 0 / 1
const addressKey1 = await deriveDogecoinAddress(1);

// Now, you can ask the user to e.g. sign transactions!
---------------------------------------------------------------------------------------
in index.ts

I'm trying to run it but it gives me an error...what am I missing?

Really really sorry for the dummy question, I hope that someone can help me.
I've looked for guides and similar discussions but I wasn't able to get it by myself (it's more than 2 days that I'm stopped here...as I said I'm a noob, unfortunately).
Thanks again

**Comment by ritave** [comment]
Hello @MrDonaldBean,

To help you out, I'd need a complete, but minimal, snap example. I also need to understand it clearly, could you please [format your question](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax) in a more understandable way? Having the code properly quoted would be very helpful.

**Comment by MrDonaldBean** [comment]
Sorry, I'll try to properly quotes.

Basically I've:

- Cloned https://github.com/MetaMask/template-snap-monorepo
- Installed @metamaks/key-tree
- Copied **code <1>** below into "template-snap-monorepo/packages/snap/snap.manifest.json"
- Copied **code <2>** below into "template-snap-monorepo/packages/snap/src/index.ts"


**code<1>:**
`{
"initialPermissions": {
"snap_getBip44Entropy": [
{
"coinType": 3
}
]
}
}`

**code<2>:**
```javascript
import { getBIP44AddressKeyDeriver } from '@metamask/key-tree';

// By way of example, we will use Dogecoin, which has coin_type 3.
const dogecoinNode = await wallet.request({
method: 'snap_getBip44Entropy',
params: {
coinType: 3,
},
});

// Next, we'll create an address key deriver function for the Dogecoin coin_type node.
// In this case, its path will be: m / 44' / 3' / 0' / 0 / address_index
const deriveDogecoinAddress = await getBIP44AddressKeyDeriver(dogecoinNode);

// These are BIP-44 nodes containing the extended private keys for
// the respective derivation paths.

// m / 44' / 3' / 0' / 0 / 0
const addressKey0 = await deriveDogecoinAddress(0);

// m / 44' / 3' / 0' / 0 / 1
const addressKey1 = await deriveDogecoinAddress(1);

// Now, you can ask the user to e.g. sign transactions!`
```

**Comment by ritave** [comment]
I see you're requesting `snap_getBip44Entropy` in a global scope. This won't work because we cancel all requests that are done outside of a callbacks such as `onRpcRequest`.

For example
```javascript
exports.onRpcRequest = async () => {
  const dogecoinNode = await wallet.request({
    method: 'snap_getBip44Entropy',
    params: { coinType: 3 },
  });
};
```

**Reply by MrDonaldBean** [reply]
I've changed "template-snap-monorepo/packages/snap/src/index.ts" into:

`// By way of example, we will use Dogecoin, which has `coin_type` 3.
exports.onRpcRequest = async () => {
  const dogecoinNode = await wallet.request({
    method: 'snap_getBip44Entropy',
    params: {
      coinType: 3,
    },
  });
};

// Next, we'll create an address key deriver function for the Dogecoin coin_type node.
// In this case, its path will be: m / 44' / 3' / 0' / 0 / address_index
const deriveDogecoinAddress = await getBIP44AddressKeyDeriver(dogecoinNode);

// These are BIP-44 nodes containing the extended private keys for
// the respective derivation paths.

// m / 44' / 3' / 0' / 0 / 0
const addressKey0 = await deriveDogecoinAddress(0);

// m / 44' / 3' / 0' / 0 / 1
const addressKey1 = await deriveDogecoinAddress(1);

// Now, you can ask the user to e.g. sign transactions!`

But it still doesn't work....it keeps giving me "unexpected token" error:
![Screenshot from 2022-11-18 15-52-04](https://user-images.githubusercontent.com/118533967/202732845-79a9a1ae-7e27-4b56-a0f9-db9d9a92c428.png)

**Reply by ritave** [reply]
You can't address variables defined in the nested scope in the global scope. This is how JavaScript, and most other languages work.

You'd have to move all your private key derivation code inside that `onRpcRequest` callback.

While I'm happy to help in any Snap related issues, I suggest you spend some time getting more advanced with JavaScript as Snaps at this point in time requires advanced understanding of JavaScript and programming languages.

**Reply by MrDonaldBean** [reply]
Of course yes, it is the first point of my to-do list...unfortunately I've an urgency with stucked tokens so I'm looking for a faster way to get access to them and I've thought that just copying this example https://docs.metamask.io/guide/snaps-rpc-api.html#snap-getbip44entropy would not be so hard to do....but unfortunately it seems that manifest and code of the guide example are wrong :-(((

![Screenshot from 2022-11-18 17-08-58](https://user-images.githubusercontent.com/118533967/202750850-0d509e92-3489-418e-810d-16d1715b6530.png)

Now I get this error but I don't want to bother you anymore...I'll try to fix it in someway

**Comment by MrDonaldBean** [comment]
This is (maybe) what I was looking for: https://github.com/metamask/test-snaps/tree/main/packages/bip44

Problem is that when I click "connect to bip-44 snap" it gives me this error: "The method "wallet_requestSnaps" does not exist / is not available."...any clues for this error?

Sorry if this is another dummy question

![image](https://user-images.githubusercontent.com/118533967/203026333-0a760598-0675-422e-9418-cd4acae6bc2c.png)

**Reply by ritave** [reply]
`test-snap` is a repository for Quality Assurance team to verify that we didn't break anything.

You shouldn't use it outside of looking for bugs in Snaps codebase, especially with real assets, **most likely your tokens will be gone forever if you try to use it**.

**Reply by ritave** [reply]
@MrDonaldBean if you need to extract some assets from your seed words, maybe you should try talking with the main MetaMask Support? I believe we have a few technical engineers there that help solve similar problems

**Reply by MrDonaldBean** [reply]
> @MrDonaldBean if you need to extract some assets from your seed words, maybe you should try talking with the main MetaMask Support? I believe we have a few technical engineers there that help solve similar problems

Thanks for the heads up.

Yes, that's my issue...I'm struggling to find a method to sign a transaction of a different derivation path wallet (different from m/44/60/...). Of course I've the seed.

I know that it is possible to perform it via Snap through snap_getBip44Entropy...it's more than 2 weeks that I'm looking for a github working template for this (it's very strange in my mind that no one has ever built a so useful repo) but without success :-((

**Reply by ritave** [reply]
Talk to MetaMask support, way smaller chance of losing your assets.

**Reply by MrDonaldBean** [reply]
> Talk to MetaMask support, way smaller chance of losing your assets.

They redirected me to discord consensys - snap section 😔

**Reply by ritave** [reply]
@MrDonaldBean I've reached out internally to the Support team with an explanation, looks like they understand and asked you to go through https://community.metamask.io/t/sign-transaction-for-a-non-m-44-60-address/23496/8 so that they can track the issue easily.

If you get any kind of reference number, you can give it to me and I'll try to push it internally.

**Reply by MrDonaldBean** [reply]
> @MrDonaldBean I've reached out internally to the Support team with an explanation, looks like they understand and asked you to go through https://community.metamask.io/t/sign-transaction-for-a-non-m-44-60-address/23496/8 so that they can track the issue easily.
> 
> If you get any kind of reference number, you can give it to me and I'll try to push it internally.

I did it and they've just opened the ticket (968340). Thank you very much, I appreciate your effort
